A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,296 @@
+    Create a Cookie from our cookie configuration;
+    Set cookie's value as encoded_address;
+    Add the cookie header to the response;
+  }
+```
+
+### LB Policy for Stateful Session Affinity
+
+After the `CookieBasedStatefulSessionFilter` has passed the `override-host`
+value to the Load Balancer as a "pick-argument", the Load Balancer uses
+this value to route the RPC appropriately. The required logic needs to be
+implemented in a new and separate LB policy at an appropriate place in the
+hierarchy.
+
+RPC load reporting happens in the `“xds_cluster_impl”` policy and we do want
+all RPCs to be included in the load reports. And this policy conditionally
+routes an RPC based on the `override-host` value if present, or else delegates
+to the currently configured LB policy. Hence the new policy needs to be just
+below the `“xds_cluster_impl”` as its child policy. Let's call this new policy
+`“override-host-experimental”`. This policy has the required subchannel and
+RPC routing logic as follows:
+
+* maintain a map of addresses (IP:port) to subchannels - let’s call this map
+  `overrideHostMap`.
+
+* whenever a new subchannel is created (as a result of an endpoint update from
+  the control plane), add a new entry to `overrideHostMap` with the subchannel
+  address(es) as the key and the subchannel as the value.
+
+    * In Java, this may result into more than one entry if the
+      createSubchannel arg’s `equivalentGroupAddress` has more than one
+      address.
+
+    * in C-core and Node, LB policies create a new subchannel for every address
+      every time there is a new address list. In the map, we will just replace
+      an existing entry with the most recent subchannel created for a given
+      address i.e. the new subchannel created will replace the older one for
+      that address.
+
+* whenever a subchannel is shut down remove the associated entries from
+  `overrideHostMap`.
+
+    * this can be achieved by wrapping a subchannel to intercept subchannel
+      shutdown and the wrapper is returned by `createSubchannel`.
+
+* the policy's subchannel picker works as follows:
+
+    * when called to pick a subchannel, it gets the value of `override-host`
+      that was passed from `CookieBasedStatefulSessionFilter` (as described
+      above).
+
+    * if the value is not present then just delegate to the child policy.
+
+    * if the value is present, then get the address in the value to look up
+      an entry in `overrideHostMap`. If no entry is found, then delegate to
+      the child policy.
+
+    * if an entry is found, get that subchannel’s connectivity state and
+      if it is not one of [`CONNECTING`, `READY`, `IDLE`] then delegate to
+      the child policy.
+
+    * otherwise return the subchannel from the entry as the pick result.
+
+Note that we create the `“override-host-experimental”` policy as the child of
+`“xds_cluster_impl”` only when the `override_host_status` config is present
```

Comment:
this seems to break the "default" case as described here: https://github.com/envoyproxy/envoy/blob/15d8b93608bc5e28569f8b042ae666a5b09b87e9/api/envoy/config/cluster/v3/cluster.proto#L615

```
If this is unset then [UNKNOWN, HEALTHY, DEGRADED] will be applied by default. If this is
 set with an empty set of statuses then host overrides will be ignored by the load balancing.
```

Reply:
Hmmm, good point. I see 2 alternatives:

- find a way to plumb the `StatefulSessionAffinityFilter` configuration (or just its presence as a flag) in LDS/RDS to gRPC's LB hierarchy and that's quite difficult or invasive
- we just update the description of the field in cluster.proto to describe how gRPC uses it. This is much easier and preferable if you think there is no major impact on users

Reply:
The second option is suboptimal as it would be causing deployment inconsistencies between regular and proxy-less setups. 

What is the overhead of having this install all the time or perhaps embed the functionality directly into the child LB policy.

Reply:
For the first point, the user can set the field to the default value `[UNKNOWN, HEALTHY, DEGRADED]` if they know the config will be used by proxyless gRPC (and Envoy) to prevent the deployment inconsistencies.

For the overhead I'll let @markdroth comment on it.

Reply:
For the first point, the inconsistency is still there and need to be heavily documented to avoid issues in mixed deployments.

Reply:
> ...the inconsistency is still there...

Inconsistency, may be yes but you can  have a config that works for both (envoy and grpc)

Reply:
> If the override_host_experimental policy is disabled, it should still maintain the map in the control plane, so that it has all of the necessary data to start working if it becomes enabled later, but it does not need to inject its own picker.

How does it "become enabled later" if the `override_host_status` unset by default and indicates `[UNKNOWN, HEALTHY, DEGRADED]` values.

Reply:
The feature can become enabled or disabled at any time when the client receives a CDS update that changes the value of `override_host_status`.  If the old value of that field did not include `UNKNOWN` or `HEALTHY` and the new value does, then the feature becomes enabled, and if the old value did include `UNKNOWN` or `HEALTHY` and the new value does not, then the feature becomes disabled.